libtracepoint-control-cpp/src/TracepointSession.cpp

```diff
@@ -748,6 +832,107 @@ TracepointSession::GetBufferFiles(
 }
 
+_Success_(return == 0) int
+TracepointSession::SavePerfDataFile(
+    _In_z_ char const* perfDataFileName,
+    int mode) noexcept
+{
+    int error = 0;
+    PerfDataFileWriter output;
+
+    error = output.Create(perfDataFileName, mode);
+    if (error != 0)
+    {
+        goto Done;
+    }
+
+    if (m_bufferLeaderFiles != nullptr)
+    {
+        IovecList vecList;
+
+        auto recordFn = [this, &vecList](
+            uint8_t const* bufferData,
+            uint16_t recordSize,
+            uint32_t recordBufferPos) noexcept
+        {
+            // Add event data to vecList.
+
+            auto const unmaskedPosEnd = recordBufferPos + recordSize;
+            if (unmaskedPosEnd <= m_bufferSize)
+            {
+                // Event does not wrap.
+                vecList.Add(bufferData + recordBufferPos, recordSize);
+            }
+            else
+            {
+                // Event wraps.
+                vecList.Add(bufferData + recordBufferPos, m_bufferSize - recordBufferPos);
+                vecList.Add(bufferData, unmaskedPosEnd - m_bufferSize);
+            }
+
+            // Look up the correct value for m_enumEventInfo.event_desc.
+
+            m_enumEventInfo.event_desc = nullptr;
+            if (PERF_RECORD_SAMPLE == BufferDataPosToHeader(bufferData, recordBufferPos)->type)
+            {
+                // TODO: We don't need a full parse here. Could potentially
+                // save a few cycles by inlining ParseSample and removing the
+                // parts we don't need.
+
+                // If this succeeds it will set m_enumEventInfo.event_desc.
+                ParseSample(bufferData, recordSize, recordBufferPos);
+            }
+            return true;
+        };
+
+        // Pause one buffer at a time.
+        for (uint32_t bufferIndex = 0; bufferIndex != m_bufferCount; bufferIndex += 1)
+        {
+            EnumeratorBegin(bufferIndex);
+
+            while (EnumeratorMoveNext(bufferIndex, recordFn))
+            {
+                if (m_enumEventInfo.event_desc)
+                {
+                    error = output.AddTracepointEventDesc(*m_enumEventInfo.event_desc);
+                    if (error != EEXIST && error != 0)
+                    {
+                        EnumeratorEnd(bufferIndex);
+                        goto Done;
+                    }
+                }
+
+                if (vecList.RoomLeft() < 2) // Next recordFn may need up to 2 calls to Add().
+                {
+                    error = vecList.Flush(output);
+                    if (error != 0)
+                    {
+                        EnumeratorEnd(bufferIndex);
+                        goto Done;
+                    }
+                }
+            }
+
+            error = vecList.Flush(output);
+
+            EnumeratorEnd(bufferIndex);
+
+            if (error != 0)
+            {
+                goto Done;
+            }
+        }
+    }
+
+    // TODO: Other headers.
```

Comment:
Note: The current code is minimal and has no optional headers, such as header with system information. Many of these optional headers are probably unnecessary if we're focused on just recording tracepoints, but a few are probably useful. I'll evaluate these and add a few of them in a subsequent PR.